File: cpp/0206反转链表.cpp
#include<vector>
#include<iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
};

ListNode *Create(vector<int> nums)
{
    if (!nums.size())
        return NULL;
    ListNode *head = new ListNode();
    ListNode *p = head;
    ListNode *q;

    p->val = nums[0];
    for (int i = 1; i < nums.size(); ++i)
    {
        ListNode *q;
        q = new ListNode();
        q->val = nums[i];
        p->next = q;
        p = q;
    }
    p->next = NULL;
    return head;
}

void Show(ListNode *head)
{
    for (; head != NULL; head = head->next)
    {
        cout << head->val << " ";
    }
}


ListNode *reverseList(ListNode *head)
{
    if(!head)
        return NULL;
    ListNode *prev = NULL;
    ListNode *cur = head;

    while(cur)
    {
        ListNode *temp = cur->next;
        cur->next = prev;
        prev = cur;
        cur = temp;
    }

    return prev;
}

File: python/0206反转链表.py
from linklist import *

# 递归
def reverseList1(head: ListNode) -> ListNode:
    if(not head):
        return None
    if(not head.next):
        return head

    reversed_h=reverseList1(head.next)
    p=reversed_h

    while(p.next):
        p=p.next

    p.next=head
    head.next=None

    return reversed_h

# 双指针
def reverseList2(head: ListNode) -> ListNode:
    if(not head):
        return None
    
    prev=None
    cur=head
    while(cur):
        temp=cur.next
        cur.next=prev
        prev=cur
        cur=temp
    
    return prev

h=create([1,2,3])

show(reverseList2(h))